src/tools/idlc/src/generator_common/common.c

```diff
@@ -0,0 +1,201 @@
+    if (off == 0)
+      break;
+    cnt = strlen(sep);
+    off -= cnt;
+    copy(str, off, size, sep, cnt);
+  }
+  str[ (size > len ? len : size - 1) ] = '\0';
+  return (int)len;
+}
+
+static int print_templ_type(
+  char *str, size_t size, const void *node, void *user_data)
+{
+  const idl_type_spec_t *type_spec;
+  char dims[32], *name = NULL;
+  const char *seg, *type;
+  size_t cnt, len = 0, seq = 0;
+
+  (void)user_data;
+  if (idl_type(node) == IDL_STRING)
+    return idl_snprintf(str, size, "char");
+  /* sequences require a little magic */
+  assert(idl_type(node) == IDL_SEQUENCE);
+
+  dims[0] = '\0';
+  type_spec = idl_type_spec(node);
+  for (; idl_is_sequence(type_spec); type_spec = idl_type_spec(type_spec))
+    seq++;
+
+  if (idl_is_base_type(type_spec) || idl_is_string(type_spec)) {
+    switch (idl_type(type_spec)) {
+      case IDL_BOOL:     type = "bool";                break;
+      case IDL_CHAR:     type = "char";                break;
+      case IDL_INT8:     type = "int8";                break;
+      case IDL_OCTET:    type = "octet";               break;
+      case IDL_UINT8:    type = "uint8";               break;
+      case IDL_SHORT:    type = "short";               break;
+      case IDL_INT16:    type = "int16";               break;
+      case IDL_USHORT:   type = "unsigned_short";      break;
+      case IDL_UINT16:   type = "uint16";              break;
+      case IDL_LONG:     type = "long";                break;
+      case IDL_INT32:    type = "int32";               break;
+      case IDL_ULONG:    type = "unsigned_long";       break;
+      case IDL_UINT32:   type = "uint32";              break;
+      case IDL_LLONG:    type = "long_long";           break;
+      case IDL_INT64:    type = "int64";               break;
+      case IDL_ULLONG:   type = "unsigned_long_long";  break;
+      case IDL_UINT64:   type = "uint64";              break;
+      case IDL_FLOAT:    type = "float";               break;
+      case IDL_DOUBLE:   type = "double";              break;
+      case IDL_LDOUBLE:  type = "long_double";         break;
+      case IDL_STRING:
+        if (idl_is_bounded(type_spec))
+          idl_snprintf(dims, sizeof(dims), "%"PRIu32, idl_bound(type_spec));
+        type = "string";
+        break;
+      default:
+        abort();
+    }
+  } else {
+    if (IDL_PRINT(&name, print_type, type_spec) < 0)
+      return -1;
+    type = name;
+  }
+
+  seg = "dds_sequence_";
+  cnt = strlen(seg);
+  copy(str, len, size, seg, cnt);
+  len += cnt;
+  seg = "sequence_";
+  cnt = strlen(seg);
+  for (; seq; len += cnt, seq--)
+    copy(str, len, size, seg, cnt);
+  cnt = strlen(type);
+  copy(str, len, size, type, cnt);
+  len += cnt;
+  cnt = strlen(dims);
+  copy(str, len, size, dims, cnt);
+  len += cnt;
+  str[ (size > len ? len : size - 1) ] = '\0';
+  if (name)
+    idl_free(name);
+  return (int)len;
+}
+
+int print_type(char *str, size_t size, const void *ptr, void *user_data)
```

Comment:
Isn't this one quite specific to C? IINM it is used only by the C backend and by the type object generator in the "print" operation, which prints the macros defining the type info and type maps as they are used by the C backend.

Of course the C++ could use those, too, but it doesn't seem to do so. I think that was the correct decision, you really don't want macros in C++ if you can use `constexpr`s instead.

So I think this one is in the wrong library. (Plus however that propagates through the code.)

Reply:
Some functions might be able to be moved from the descriptor_type_meta to other (more limited) scopes.
I will look in to this.